src/tools/wasm-opt.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2015 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// A WebAssembly optimizer, loads code, optionally runs passes on it,
+// then writes it.
+//
+
+#include <memory>
+
+#include "pass.h"
+#include "support/command-line.h"
+#include "support/file.h"
+#include "wasm-printing.h"
+#include "wasm-s-parser.h"
+#include "wasm-validator.h"
+
+using namespace wasm;
+
+//
+// main
+//
+
+int main(int argc, const char* argv[]) {
+  Name entry;
+  std::vector<std::string> passes;
+
+  Options options("wasm-opt", "Optimize .wast files");
+  options
+      .add("--output", "-o", "Output file (stdout if not specified)",
+           Options::Arguments::One,
+           [](Options* o, const std::string& argument) {
+             o->extra["output"] = argument;
+             Colors::disable();
+           })
+      .add("", "-O", "execute default optimization passes",
+           Options::Arguments::Zero,
+           [&passes](Options*, const std::string&) {
+             passes.push_back("O");
+           })
+      .add_positional("INFILE", Options::Arguments::One,
+                      [](Options* o, const std::string& argument) {
+                        o->extra["infile"] = argument;
+                      });
+  for (const auto& p : PassRegistry::get()->getRegisteredNames()) {
+    options.add(
+        std::string("--") + p, "", PassRegistry::get()->getPassDescription(p),
+        Options::Arguments::Zero,
+        [&passes, p](Options*, const std::string&) { passes.push_back(p); });
+  }
+  options.parse(argc, argv);
+
+  auto input(read_file<std::string>(options.extra["infile"], Flags::Text, options.debug ? Flags::Debug : Flags::Release));
+
+  Module wasm;
+
+  try {
+    if (options.debug) std::cerr << "s-parsing..." << std::endl;
+    SExpressionParser parser(const_cast<char*>(input.c_str()));
+    Element& root = *parser.root;
+    if (options.debug) std::cerr << "w-parsing..." << std::endl;
+    SExpressionWasmBuilder builder(wasm, *root[0]);
+    assert(WasmValidator().validate(wasm));
```

Comment:
Since it's possible for input to be syntactically correct (parseable) but invalid, this should probably be a `Fatal()` instead of an assert. And probably we want it to run even for release builds, so it doesn't confuse the passes, no?